File: Server/src/packet.py
import typing

from protos.packetCode_pb2 import PacketCode

PACKET_CONST_HEAD = 0x2B3C
PACKET_CONST_TAIL = 0x4D5F

def _readUInt16(data: bytes, startIndex: int) -> int:
    b1 = data[startIndex]
    b2 = data[startIndex + 1]
    return (b1 << 8) | b2

def _writeUInt16(buffer: bytearray, value: int) -> None:
    buffer.append((value >> 8) & 0xFF)
    buffer.append(value & 0xFF)

class Packet(object):
    def __init__(self, packetCode: PacketCode, payloadBytes: bytes | bytearray | None = None) -> None:
        self.packetCode = packetCode
        self.payloadBytes = payloadBytes

    def encode(self) -> bytearray:
        buffer = bytearray()
        _writeUInt16(buffer, PACKET_CONST_HEAD)
        _writeUInt16(buffer, self.packetCode)

        if self.payloadBytes is None:
            _writeUInt16(buffer, 0)
        else:
            _writeUInt16(buffer, len(self.payloadBytes))
            buffer.extend(self.payloadBytes)

        _writeUInt16(buffer, PACKET_CONST_TAIL)
        return buffer

    @classmethod
    def decode(cls, data: bytes) -> typing.Union["Packet", None]:
        # sizeof(CONST_HEAD + PacketCode + PayloadLength + CONST_TAIL) == 8
        if len(data) < 8:
            return None

        # Check Head Const
        if _readUInt16(data, 0) != PACKET_CONST_HEAD:
            return None

        payloadLength = _readUInt16(data, 4)

        # Check Packet Size
        if len(data) != 8 + payloadLength:
            return None

        # Check Tail Const
        if _readUInt16(data, 6 + payloadLength) != PACKET_CONST_TAIL:
            return None

        packetCode = _readUInt16(data, 2)
        payloadBytes = data[6:6+payloadLength]
        return cls(packetCode, payloadBytes)

File: Client/StarRailMotionCapture/Assets/Scripts/Runtime/Net/Packet.cs
using HSR.MotionCapture.Net.Protos;
using JetBrains.Annotations;
using System;

namespace HSR.MotionCapture.Net
{
    public static class Packet
    {
        public delegate int WritePayloadAction<in T>(Span<byte> span, T payload);

        public const ushort CONST_HEAD = 0x2B3C;
        public const ushort CONST_TAIL = 0x4D5F;

        public static bool TryRead(Span<byte> bytes, out PacketCode code, out ReadOnlySpan<byte> payloadBytes)
        {
            code = PacketCode.None;
            payloadBytes = ReadOnlySpan<byte>.Empty;

            // sizeof(CONST_HEAD + PacketCode + PayloadLength + CONST_TAIL) == 8
            if (bytes.Length < 8)
            {
                return false;
            }

            // Check Head Const
            if (ReadUInt16(bytes[..2]) != CONST_HEAD)
            {
                return false;
            }

            int payloadLength = ReadUInt16(bytes[4..6]);

            // Check Packet Size
            if (bytes.Length != 8 + payloadLength)
            {
                return false;
            }

            // Check Tail Const
            if (ReadUInt16(bytes[^2..]) != CONST_TAIL)
            {
                return false;
            }

            code = (PacketCode)ReadUInt16(bytes[2..4]);
            payloadBytes = bytes.Slice(6, payloadLength);
            return true;
        }

        public static int Write<T>(Span<byte> buffer, PacketCode packetCode, T payload, [CanBeNull] WritePayloadAction<T> writePayloadBytes)
        {
            WriteUInt16(buffer[..2], CONST_HEAD);
            WriteUInt16(buffer[2..4], (ushort)packetCode);
            int payloadLength = writePayloadBytes?.Invoke(buffer[6..], payload) ?? 0;
            WriteUInt16(buffer[4..6], (ushort)payloadLength);
            WriteUInt16(buffer.Slice(6 + payloadLength, 2), CONST_TAIL);
            return 8 + payloadLength;
        }

        private static ushort ReadUInt16(Span<byte> bytes)
        {
            return (ushort)((bytes[0] << 8) | bytes[1]);
        }

        private static void WriteUInt16(Span<byte> buffer, ushort value)
        {
            buffer[0] = (byte)((value >> 8) & 0xFF);
            buffer[1] = (byte)(value & 0xFF);
        }
    }
}
